packages/block-library/src/template-part/edit.js

```diff
@@ -1,3 +1,118 @@
-export default function TemplatePartEdit() {
-	return 'Template Part Placeholder';
+/**
+ * WordPress dependencies
+ */
+import {
+	useEntityProp,
+	__experimentalUseEntitySaving,
+	EntityProvider,
+} from '@wordpress/core-data';
+import { useMemo, useCallback } from '@wordpress/element';
+import { parse } from '@wordpress/blocks';
+import { serializeBlocks } from '@wordpress/editor';
+import { Button } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { InnerBlocks } from '@wordpress/block-editor';
+import { useSelect } from '@wordpress/data';
+
+const saveProps = [ 'content', 'status' ];
+function TemplatePart() {
+	const [ content, _setContent ] = useEntityProp(
+		'postType',
+		'wp_template_part',
+		'content'
+	);
+	const [ status, setStatus ] = useEntityProp(
+		'postType',
+		'wp_template_part',
+		'status'
+	);
+	const initialBlocks = useMemo( () => {
+		if ( status !== 'publish' ) {
+			// Publish if still an auto-draft.
+			setStatus( 'publish' );
+		}
+		if ( typeof content !== 'function' ) {
+			const parsedContent = parse( content );
+			return parsedContent.length ? parsedContent : undefined;
+		}
+	}, [] );
+	const [ blocks = initialBlocks, setBlocks ] = useEntityProp(
+		'postType',
+		'wp_template_part',
+		'blocks'
+	);
+	const [ isDirty, isSaving, save ] = __experimentalUseEntitySaving(
+		'postType',
+		'wp_template_part',
+		saveProps
+	);
+	const saveContent = useCallback( () => {
+		_setContent( content( { blocks } ) );
+		save();
+	}, [ content, blocks ] );
+	const setContent = useCallback( () => {
+		_setContent( ( { blocks: blocksForSerialization = [] } ) =>
+			serializeBlocks( blocksForSerialization )
+		);
+	}, [] );
+	return (
+		<>
+			<Button
+				isPrimary
+				className="wp-block-template-part__save-button"
+				disabled={ ! isDirty || ! content }
+				isBusy={ isSaving }
+				onClick={ saveContent }
+			>
+				{ __( 'Save' ) }
+			</Button>
+			<InnerBlocks value={ blocks } onChange={ setBlocks } onInput={ setContent } />
```

Comment:
These new props `value` and `onChange` (_aside: also not documented_) seem contrary to the purpose of `InnerBlocks` as an abstracted, managed rendering of blocks. For something like this where we want full control over the rendering of blocks, is there any reason not to render a custom `BlockEditorProvider` instead?

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/provider/README.md

Or, alternatively, based on what we're doing above with serializing blocks and how it's proposed to expose the internal state utilities, maybe this is something which should be baked into `InnerBlocks`. Is the idea that we need some way to customize the "save" or "change" behaviors, in this case to sync to the entity?

For that, I might be okay with an `onChange`, but more as a true side-effect, where we're not intending to fully control the value of `InnerBlocks`, but rather sync any changes.

Then again, that seems like something we could also achieve without a callback, and instead using the `getBlocks` selector to track changes to the inner blocks of this block. 

To me, this is a very similar problem to what was previously discussed in #5596 around having a callback to `setAttributes`, where instead the model we encourage is where one should detect the changes using the existing selectors functionality (even easier now today with hooks than it was at the time of #5596).

Reply:
>is there any reason not to render a custom `BlockEditorProvider` instead?

We don't want a split block list. We want the child blocks to still save to the parent through the block's `save` component where it makes sense, and we want the UX of a single block list.

This change provides a way to provide an initial value and then sync with the changes of `InnerBlocks` so that other code can use the child blocks to do things like save to other entities.

We could achieve something similar with actions and selectors, albeit in a much more verbose and error-prone way. We would probably end up implementing a custom hook for it and the logic for checking if the last check was persistent might get messy. This approach would also be less performant.

I don't think this is comparable to a callback for `setAttributes`. In that scenario people needed a way to react to changes in the store, and that's what subscriptions are for. In this scenario we want to control/sync with the state of a component.

What are the issues you see this creating in the future?

Reply:
>Maybe if we compare it to React's distinction of controlled vs. uncontrolled inputs, it's fine to consider support for both use-cases.

Yes, I think so.

>are we serializing blocks on every keypress within a template?

No, we mirror what the editor does and only do it on save.

>I'm not sure it's something we need to act on now, but it gets me me wondering whether there are alternative / additional abstractions we could imagine to help encapsulate this behavior of mapping the blocks data to a custom entity serialization. As before I was talking of this sort of InnerBlocks being an enhanced version of a BlockEditorProvider, I wonder if there's some parallel of how Editor is an enhanced version of BlockEditor responsible for managing how a block editor saves to a post.

I have some ideas to make this better and easier to understand with a custom hook. I'll also try the selector approach to see how viable it is and if there is a noticeable performance drop.

Reply:
>I think the issue for me is that we've now introduced (to me at least) some confusion around exactly what InnerBlocks is. Is it a controlled input? Or is it abstracted to maintain the value internally? Where before it was clearly the latter, with these changes it seems like the answer is now "both". A concern stems from the need to document that clearly, and to maintain that in perpetuity.

That's fair. I agree we should avoid it if possible.

>For me, this sort of controlled rendering was was BlockEditor should do. I understand the usability need that we need blocks within this get treated as part of the same "tree". It reminds me of similar work in #14715 with the reusable block, and where we want to draw this line of separate vs. integrated editors. Is there value in the reusability of block-editor if these limitations prevent us from using it in the editor? Is the purpose of these InnerBlocks revisions to serve as an enhanced version of a block editor? In the end, maybe this is the right change, since we could have a use case for each, but it is worth having the clear picture for how all these fit together.

`BlockEditor` is for building entire new editors. `InnerBlocks` is for managing subtrees in editors.

>Maybe if we compare it to React's distinction of controlled vs. uncontrolled inputs, it's fine to consider support for both use-cases. I think there's a bit of a difference in how what would become the equivalent of an "uncontrolled" InnerBlocks has some inherited behavior (and an inherited default value of an inner block) whereas, without a defaultValue, any other input would always be empty.

Yes, that's what I based it on.

>To a lesser degree, I still think we'll have some concern about how difficult it is to manage this value from a block. For example, with this implementation, are we serializing blocks on every keypress within a template? That seems like it could become a performance concern for non-trivial templates. Whether that's something we could improve, it would be fair to acknowledge that others may encounter similar pain points.

We're not serializing on every keypress, but I also see this becoming a common pattern and it would be good to deal with all the pain points in one place.

>I'm not sure it's something we need to act on now, but it gets me me wondering whether there are alternative / additional abstractions we could imagine to help encapsulate this behavior of mapping the blocks data to a custom entity serialization. As before I was talking of this sort of InnerBlocks being an enhanced version of a BlockEditorProvider, I wonder if there's some parallel of how Editor is an enhanced version of BlockEditor responsible for managing how a block editor saves to a post.

We could have an enhanced `InnerBlocks` that lets you turn on the syncing with a prop, but that would make `block-editor` depend on `core-data`. I think a `useSyncedEntityInnerBlocks` hook will work here. I've thought a lot about this and have much better ideas for how to deal with all this in a better way, specially with the new changes to `master`. I'll close this PR and open a new one soon.